src/ray/rpc/server_call.h

```diff
@@ -20,72 +21,111 @@ namespace rpc {
 using SendReplyCallback = std::function<void(Status status, std::function<void()> success,
                                              std::function<void()> failure)>;
 
+/// At present, this gRPC server supports two types of request.
+/// The first is default asynchronous request call that each call contains exactly a
+/// request and a reply. Second, asynchronous stream call that the client can send a
+/// sequence of request messages and receive a sequence of reply messages.
+/// The main difference between them is that the requests sent by a stream call are kept
+/// in order and not all the requests should be with a corresponding reply, it's totally
+/// determined by user.
+enum class ServerCallType {
+  DEFAULT_ASYNC_CALL,
+  STREAM_ASYNC_CALL,
+};
+
 /// Represents state of a `ServerCall`.
 enum class ServerCallState {
+  CONNECT,
   /// The call is created and waiting for an incoming request.
   PENDING,
   /// Request is received and being processed.
   PROCESSING,
   /// Request processing is done, and reply is being sent to client.
-  SENDING_REPLY
+  SENDING_REPLY,
+  /// Stream has finished.
+  FINISH,
 };
 
 class ServerCallFactory;
-
+class ServerCallTag;
 /// Represents an incoming request of a gRPC server.
 ///
-/// The lifecycle and state transition of a `ServerCall` is as follows:
-///
-/// --(1)--> PENDING --(2)--> PROCESSING --(3)--> SENDING_REPLY --(4)--> [FINISHED]
-///
-/// (1) The `GrpcServer` creates a `ServerCall` and use it as the tag to accept requests
-///     gRPC `CompletionQueue`. Now the state is `PENDING`.
-/// (2) When a request is received, an event will be gotten from the `CompletionQueue`.
-///     `GrpcServer` then should change `ServerCall`'s state to PROCESSING and call
-///     `ServerCall::HandleRequest`.
-/// (3) When the `ServiceHandler` finishes handling the request, `ServerCallImpl::Finish`
-///     will be called, and the state becomes `SENDING_REPLY`.
-/// (4) When the reply is sent, an event will be gotten from the `CompletionQueue`.
-///     `GrpcServer` will then delete this call.
-///
 /// NOTE(hchen): Compared to `ServerCallImpl`, this abstract interface doesn't use
 /// template. This allows the users (e.g., `GrpcServer`) not having to use
 /// template as well.
 class ServerCall {
  public:
-  /// Get the state of this `ServerCall`.
-  virtual ServerCallState GetState() const = 0;
-
-  /// Set state of this `ServerCall`.
-  virtual void SetState(const ServerCallState &new_state) = 0;
-
   /// Handle the requst. This is the callback function to be called by
   /// `GrpcServer` when the request is received.
   virtual void HandleRequest() = 0;
 
   /// Get the factory that created this `ServerCall`.
-  virtual const ServerCallFactory &GetFactory() const = 0;
+  virtual const ServerCallFactory &GetFactory() const {}
```

Comment:
I think it would be better to put these methods into corresponding derived classes, and perform a downcast based on ServerCallType when calling these functions.

Reply:
same for all the methods above that are not common, but specific to one of the specific ServerCall.

Reply:
Currently the superclasses ServerCall or ClientCall are used in many places, should I change all the definition to the implementation subclasses? I want more comments on it.

Reply:
@raulchen @robertnishihara @pcmoritz  what do you guys think?

Reply:
Agreed that stream-specific methods should be put in the `ServerStreamCall` class. 
I think we can
1) put `ServerCallType` in the tag class.
2) downcast `tag.call_` to the specific sub class type based on the type, and pass it to `ProcessStreamCall` or `ProcessDefaultCall`